### PR TITLE
JAMES-2632 Optimize List all mailboxes

### PR DESCRIPTION
**Pull request by chibenwa**
Follow up of https://github.com/linagora/james-project/pull/2770

### CODE REVIEW

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 1]
```diff
@@ -131,9 +134,34 @@ public long countUnseenMessagesInMailbox(Mailbox mailbox) throws MailboxExceptio
 
     @Override
     public MailboxCounters getMailboxCounters(Mailbox mailbox) throws MailboxException {
+<<<<<<< HEAD
```
Comment (rouazana) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: rebase issue ?^^

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 2]
```diff
@@ -161,7 +152,6 @@ public MailboxCounters getMailboxCounters(Mailbox mailbox) throws MailboxExcepti
             .concatMap(this::getMailboxCounters)
             .toStream()
             .collect(Guavate.toImmutableList());
->>>>>>> f4ef8da35d... JAMES-2632 Allow bulk mailbox counter retrieval
```
Comment (rouazana) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: eurk, can you fix this before please?

**Comment by chibenwa** [comment]
Before 

Capture d’écran de 2019-11-07 17-36-43

After:

Capture d’écran de 2019-11-07 17-36-48

With contat map on search (last commit) :

Capture d’écran de 2019-11-07 17-40-12

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapper.java [hunk 3]
```diff
@@ -155,7 +155,7 @@ private Mailbox addAcl(MailboxACL acl, Mailbox mailbox) {
 
         return listUserMailboxes
                 .filter(idAndPath -> regex.matcher(idAndPath.getMailboxPath().getName()).matches())
-                .flatMap(this::retrieveMailbox)
+                .concatMap(this::retrieveMailbox)
```
Comment (blackheaven) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapper.java]: You may slow down on retreive

**Comment by chibenwa** [comment]
```

[0414fa0ed756b10915d071f5218b2f74cdd4c786] [ERROR]   MailboxManagerTest$SearchTests.searchShouldReturnACL:916 [List check first element] 
[0414fa0ed756b10915d071f5218b2f74cdd4c786] Expecting:
[0414fa0ed756b10915d071f5218b2f74cdd4c786]  <MailboxACL{entries={owner=aeiklprstwx}}>
[0414fa0ed756b10915d071f5218b2f74cdd4c786] to be equal to:
[0414fa0ed756b10915d071f5218b2f74cdd4c786]  <MailboxACL{entries={USER_2=lr, owner=aeiklprstwx}}>
[0414fa0ed756b10915d071f5218b2f74cdd4c786] but was not.
[0414fa0ed756b10915d071f5218b2f74cdd4c786] [ERROR] Errors: 
[0414fa0ed756b10915d071f5218b2f74cdd4c786] [ERROR]   JpaMessageMapperTest>MessageMapperTest.getMailboxCountersShouldReturnStoredValue:132 » NotImplemented
```

File: server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/model/MailboxFactory.java [hunk 4]
```diff
@@ -86,18 +98,55 @@ public MailboxBuilder usingPreloadedMailboxesMetadata(Optional<List<MailboxMetaD
...
+                MailboxId mailboxId = computeMailboxId();
+                Mono<MessageManager> mailbox = mailbox(mailboxId).cache();
+
+                MailboxACL mailboxACL = mailboxMetaData.map(MailboxMetaData::getResolvedAcls)
+                    .orElseGet(Throwing.supplier(() -> retrieveCachedMailbox(mailboxId, mailbox).getResolvedAcl(session)).sneakyThrow());
+
+                MailboxPath mailboxPath = mailboxMetaData.map(MailboxMetaData::getPath)
+                    .orElseGet(Throwing.supplier(() -> retrieveCachedMailbox(mailboxId, mailbox).getMailboxPath()).sneakyThrow());
+
+                MailboxCounters mailboxCounters = mailboxMetaData.map(MailboxMetaData::getCounters)
```
Comment (remk) [inline server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/model/MailboxFactory.java]: Can you extract the method to retrieve the mailboxACL, mailboxPath, mailboxCounters from the metadata?

Comment (chibenwa) [reply server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/model/MailboxFactory.java]: I dont understand what you ask for...

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
`[ERROR] The forked VM terminated without properly saying goodbye. VM crash or System.exit called?`

test this please

**Comment by chibenwa** [comment]
test this pleae

**Comment by chibenwa** [comment]
test this please

File: server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/model/mailbox/MailboxNamespace.java [hunk 5]
```diff
@@ -42,6 +42,7 @@
     }
 
     public static MailboxNamespace delegated(String owner) {
+        Preconditions.checkArgument(owner != null);
```
Comment (blackheaven) [inline server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/model/mailbox/MailboxNamespace.java]: `StringUtils.isNullOrEmpty`?

Comment (chibenwa) [reply server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/model/mailbox/MailboxNamespace.java]: Enpty use case is handled the line below by "isBlank" check...

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMailboxMapper.java [hunk 6]
```diff
@@ -207,22 +198,6 @@ public void delete(Mailbox mailbox) throws MailboxException {
...
-        try {
-            getEntityManager().createNamedQuery("deleteAllMemberships").executeUpdate();
-        } catch (PersistenceException e) {
-            throw new MailboxException("Delete of mailboxes failed", e);
-        } 
-    }
-    
-    public void deleteAllMailboxes() throws MailboxException {
-        try {
-            getEntityManager().createNamedQuery("deleteAllMailboxes").executeUpdate();
```
Comment (rouazana) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMailboxMapper.java]: you can also remove "deleteAllMailboxes" in mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/model/JPAMailbox.java

File: server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/model/mailbox/MailboxNamespace.java (hunk 5 above)
Comment (blackheaven) [reply server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/model/mailbox/MailboxNamespace.java]: sure, but why not doing the two checks at the same time?

File: server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/model/MailboxFactory.java [hunk 7]
```diff
@@ -86,18 +98,55 @@ public MailboxBuilder usingPreloadedMailboxesMetadata(Optional<List<MailboxMetaD
...
             } catch (MailboxNotFoundException e) {
                 return Optional.empty();
             } catch (MailboxException e) {
                 throw new RuntimeException(e);
             }
         }
+
+        private MailboxId computeMailboxId() {
+            int idCount = Booleans.countTrue(id.isPresent(), mailboxMetaData.isPresent());
+            Preconditions.checkState(idCount == 1, "You need exectly one 'id' 'mailboxMetaData'");
```
Comment (remk) [inline server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/model/MailboxFactory.java]: s/exectly/exactly

**Comment by chibenwa** [comment]
test this please

**Comment by Arsnael** [comment]
```
[ERROR]   ConcurrentTestRunnerTest.runShouldPerformAllOperationsEvenOnOccasionalExceptions:230 » NotTerminated
```
test this please

**Comment by chibenwa** [comment]
Just rebased to solve a minor import conflict.

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**